File: reference/rocketride-server/packages/client-python/src/rocketride/core/dap_client.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Debug Adapter Protocol (DAP) Client Implementation.

This module provides the DAP client implementation that handles communication
with RocketRide servers using the Debug Adapter Protocol. It manages connection
lifecycle, request/response correlation, and message handling.

This is an internal implementation detail that supports the higher-level
RocketRide client functionality. Most users won't need to interact with this
directly - it's used internally by RocketRideClient for server communication.

Key Features:
- WebSocket-based communication with servers
- Request/response correlation using sequence numbers
- Connection lifecycle management
- Message serialization and deserialization
- Error handling and recovery
"""

import asyncio
from typing import Dict, Any, Optional, TYPE_CHECKING
from .dap_base import DAPBase

if TYPE_CHECKING:
    from ..types.client import ConnectResult


class DAPClient(DAPBase):
    """
    DAP (Debug Adapter Protocol) client for communicating with RocketRide servers.

    This class implements the client side of the DAP communication protocol,
    handling connection management, message correlation, and protocol compliance.
    It provides the underlying communication infrastructure used by the
    higher-level RocketRide client functionality.

    Key Responsibilities:
    - Establishing and maintaining connections to DAP servers
    - Correlating requests with their responses using sequence numbers
    - Handling connection events and state changes
    - Managing the message sending and receiving lifecycle
    - Providing error handling and recovery mechanisms

    This is used internally by RocketRideClient and its mixins to communicate
    with RocketRide servers reliably and efficiently.
    """

    def __init__(self, **kwargs) -> None:
        """
        Initialize the DAP client infrastructure.

        Sets up the client communication system, initializes request/response
        correlation, and prepares connection state management.

        Args:
            **kwargs: Configuration parameters including transport setup
        """
        # Initialize parent class for DAP functionality
        super().__init__(**kwargs)

        # Request/response correlation system for async operations
        self._pending_requests: Dict[int, asyncio.Future] = {}
        # Default request timeout in ms (None = no timeout); converted to seconds for asyncio
        self._request_timeout: Optional[float] = kwargs.get('request_timeout', None)
        # Auth response body from the last successful connect. Set before on_connected fires.
        self._connect_result: Optional['ConnectResult'] = None

    async def _send(self, message: Dict[str, Any]) -> None:
        """
        Send a message to the DAP server through the transport layer.

        Args:
            message: The DAP message to send
        """
        return await self._transport.send(message)

    async def on_connected(self, connection_info: Optional[str] = None) -> None:
        """
        Handle connection established event.

        Called when the transport layer successfully connects to the server.
        Updates connection state and notifies parent classes.

        Args:
            connection_info: Optional connection details from transport
        """
        await super().on_connected(connection_info)

    def is_connected(self) -> bool:
        """
        Check whether the transport is connected.
        Backward compatible — returns True when the WebSocket is open.
        """
        if self._transport is None:
            return False
        return self._transport.is_connected()

    async def on_disconnected(self, reason: Optional[str] = None, has_error: bool = False) -> None:
        """
        Handle disconnection event and clean up pending requests.

        Called when the transport layer disconnects from the server.
        Fails all pending requests and cleans up connection state.

        Args:
            reason: Optional reason for disconnection
            has_error: Whether disconnection was due to error
        """
        # Fail all pending requests since connection is lost
        connection_error = ConnectionError(reason)

        for seq, future in self._pending_requests.items():
            if not future.done():
                future.set_exception(connection_error)

        # Clear pending requests
        self._pending_requests.clear()

        # Notify parent classes
        await super().on_disconnected(reason, has_error)

    async def on_connect_error(self, error: Exception) -> None:
        """
        Handle connection attempt failure.

        Forwards to parent classes in the mixin chain.
        """
        await super().on_connect_error(error)

    async def on_receive(self, message: Dict[str, Any]) -> None:
        """
        Handle received messages from the transport layer.

        Processes incoming DAP messages, correlating responses with their
        originating requests and forwarding events to handlers.

        Args:
            message: Complete DAP message received from server
        """
        message_type = message.get('type', '')

        if message_type == 'response':
            # Handle response correlation
            request_seq = message.get('request_seq')

            if request_seq is not None and request_seq in self._pending_requests:
                future = self._pending_requests.pop(request_seq)

                if not future.done():
                    future.set_result(message)
                else:
                    self.debug_message(f'Response received for already completed request: {message}')

        elif message_type == 'event':
            # Forward events to event handler
            await self.on_event(message)

        else:
            # Unknown message type
            self.debug_message(f'Unhandled message type: {message}')

    async def request(self, message: Dict[str, Any], timeout: Optional[float] = None) -> Dict[str, Any]:
        """
        Send a request and wait for the corresponding response.

        Provides synchronous-style request/response communication over async
        transport by correlating requests and responses using sequence numbers.

        Args:
            message: Request message with required DAP fields
            timeout: Optional per-request timeout in ms. Overrides the default request_timeout.

        Returns:
            dict: Response message from the server

        Raises:
            RuntimeError: If request is missing required fields or fails
            ConnectionError: If connection fails during request
            asyncio.TimeoutError: If response times out
        """
        # Validate required fields
        if 'command' not in message:
            self.raise_exception(RuntimeError("Request message must include a 'command' field"))

        if 'type' not in message:
            self.raise_exception(RuntimeError("Request message must include a 'type' field"))

        if 'seq' not in message:
            self.raise_exception(RuntimeError("Request message must include a 'seq' field"))

        # Verify connection
        if not self._transport.is_connected():
            self.raise_exception(RuntimeError('Server is not connected'))

        # Get sequence number for correlation
        seq = message['seq']

        # Create future for response correlation
        future = asyncio.Future()
        self._pending_requests[seq] = future

        try:
            # Send request through transport
            await self._send(message)
        except Exception:
            self._pending_requests.pop(seq, None)
            self.raise_exception(ConnectionError('Could not send request'))

        try:
            # Wait for correlated response, with optional timeout (ms → seconds)
            effective_timeout = timeout if timeout is not None else self._request_timeout
            if effective_timeout is not None:
                response = await asyncio.wait_for(future, timeout=effective_timeout / 1000.0)
            else:
                response = await future
            return response

        except (ConnectionError, ConnectionResetError) as e:
            # Schedule transport teardown without awaiting it. Awaiting self.disconnect()
            # here forms a cycle when request() runs inside a task that is itself a
            # member of TransportWebSocket._message_tasks: the drain gather closes back
            # on the calling task, and asyncio's cancel propagation recurses on
            # _GatheringFuture.cancel until the recursion limit fires.
            self._transport.disconnect()
            self.raise_exception(ConnectionError(f'{e}'))

        except asyncio.TimeoutError:
            # Handle timeouts specifically
            self.raise_exception(asyncio.TimeoutError('Connection time out'))

        except Exception as e:
            # Handle other errors
            self.raise_exception(RuntimeError(f'Request failed: {e}'))

        finally:
            # Always clean up the pending request
            self._pending_requests.pop(seq, None)

    async def dap_request(
        self,
        command: str,
        arguments: Optional[Dict[str, Any]] = None,
        token: Optional[str] = None,
        timeout: Optional[float] = None,
    ) -> Dict[str, Any]:
        """
        Send an arbitrary DAP command with command name, arguments, and optional token.

        Convenience method that builds the request and sends it in one call.
        Equivalent to ``build_request(command, ...)`` + ``request(...)``.

        Args:
            command: The DAP command name (e.g. 'rrext_services', 'rrext_monitor', 'initialize').
            arguments: Optional arguments for the command.
            token: Optional task/session token.
            timeout: Optional per-request timeout in ms. Overrides the default request_timeout.

        Returns:
            The response message (DAP dict) from the server.

        Example:
            response = await client.dap_request('rrext_services', {}, timeout=5000)
            if client.did_fail(response):
                raise RuntimeError(response.get('message', 'Request failed'))
        """
        request = self.build_request(
            command,
            token=token,
            arguments=arguments or {},
        )
        return await self.request(request, timeout=timeout)

    async def connect(self, timeout: Optional[float] = None) -> None:
        """
        Open the transport (WebSocket) without sending any authentication.
        Authentication is handled at the RocketRideClient level via login().

        Args:
            timeout: Optional timeout in milliseconds for the WebSocket handshake.

        Raises:
            ConnectionError: If connection fails.
            TimeoutError: If connection times out.
        """
        await self._transport.connect(timeout)

    async def disconnect(self) -> None:
        """
        Close connection to the DAP server and clean up resources.

        Gracefully closes the transport connection and cleans up any
        remaining connection state or pending operations.
        """
        # Close transport connection
        await self._transport.disconnect()

File: reference/rocketride-server/packages/client-typescript/src/client/core/DAPClient.ts
/**
 * MIT License
 *
 * Copyright (c) 2026 Aparavi Software AG
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { DAPBase } from './DAPBase.js';
import { DAPMessage, RocketRideClientConfig, ConnectResult } from '../types/index.js';
import { TransportBase } from './TransportBase.js';
import { SDK_VERSION } from '../constants.js';

/**
 * DAP (Debug Adapter Protocol) client for communicating with RocketRide servers.
 *
 * This class implements the client side of the DAP communication protocol,
 * managing request/response correlation, pending request tracking, and message
 * sequencing. It extends DAPBase to inherit common protocol functionality.
 *
 * Key responsibilities:
 * - Request/response correlation via sequence numbers
 * - Pending request management with promises
 * - Connection lifecycle management
 * - Message routing and handling
 *
 * @extends DAPBase
 */
export class DAPClient extends DAPBase {
	/**
	 * Map of outstanding requests awaiting a server response.
	 * Keyed by the message sequence number assigned at send time.
	 * Each entry holds the Promise resolve/reject callbacks and an optional
	 * timeout handle so the request can be cancelled if the server is silent.
	 */
	private _pendingRequests = new Map<
		number,
		{
			resolve: (value: DAPMessage) => void;
			reject: (reason: unknown) => void;
			timer?: ReturnType<typeof setTimeout>;
		}
	>();

	/**
	 * Monotonically increasing counter used to assign unique sequence numbers
	 * to every outgoing request. Responses echo this number in `request_seq`
	 * so they can be matched back to the originating request.
	 */
	private _sequenceNumber = 0;

	/**
	 * Default timeout in milliseconds applied to every request unless overridden
	 * by the per-call `timeout` parameter. `undefined` means no timeout.
	 */
	protected _requestTimeout?: number;

	/**
	 * Human-readable name of this client (e.g. "VS Code", "TypeScript SDK").
	 * Sent to the server during authentication so operators can identify callers.
	 */
	protected _clientDisplayName?: string;

	/**
	 * Version string of this client (e.g. "0.9.4").
	 * Sent alongside `_clientDisplayName` during authentication.
	 */
	protected _clientDisplayVersion?: string;

	/** Auth response body from the last successful _dapConnect. Set before onConnected fires. */
	protected _connectResult?: ConnectResult;

	/**
	 * Creates a DAPClient instance and configures request-timeout and client
	 * identity fields from the provided config object.
	 *
	 * @param module - Logical name for this client instance (used in debug messages)
	 * @param transport - Optional pre-built transport; may be provided later via _bindTransport
	 * @param config - Client configuration options (timeout, display name/version)
	 */
	constructor(module: string, transport: TransportBase | undefined, config: RocketRideClientConfig = {}) {
		// Delegate module name, transport, and shared config to DAPBase
		super(module, transport, config);

		// Capture the default per-request timeout (undefined = no timeout)
		this._requestTimeout = config.requestTimeout;

		// Resolve the display name the client will advertise to the server during auth
		this._clientDisplayName = config.clientName || 'TypeScript SDK';

		// Resolve the version string the client will advertise to the server during auth
		this._clientDisplayVersion = config.clientVersion || SDK_VERSION;
	}

	/**
	 * Generate the next sequence number for outgoing requests.
	 *
	 * Increments the internal counter and returns the new value.
	 * The counter never resets within a client lifetime, ensuring
	 * uniqueness across all requests on a single connection.
	 */
	private getNextSeq(): number {
		// Pre-increment so seq numbers start at 1, not 0 (0 is the unset sentinel)
		return ++this._sequenceNumber;
	}

	/**
	 * Send a message to the DAP server through the transport layer.
	 *
	 * Delegates directly to the active transport's `send` method.
	 * Throws if no transport is currently bound.
	 *
	 * @param message - The fully assembled DAP message to transmit
	 * @throws Error if no transport is available (client not connected)
	 */
	protected async _send(message: DAPMessage): Promise<void> {
		// Guard: a transport must be present before any message can be sent
		if (!this._transport) {
			throw new Error('Transport not available');
		}

		// Delegate serialisation and transmission to the transport implementation
		return this._transport.send(message);
	}

	/**
	 * Handle connection established event.
	 *
	 * Chains to the parent DAPBase implementation which clears any
	 * internal error state and invokes registered connection callbacks.
	 *
	 * @param connectionInfo - Optional human-readable description of the connection
	 */
	async onConnected(connectionInfo?: string): Promise<void> {
		// Propagate to DAPBase so shared housekeeping runs
		await super.onConnected(connectionInfo);
	}

	/**
	 * Handle connection attempt failure.
	 *
	 * Chains to the parent DAPBase implementation for shared error handling.
	 *
	 * @param error - The error that caused the connection attempt to fail
	 */
	async onConnectError(error: Error): Promise<void> {
		// Propagate to DAPBase for centralised error handling
		await super.onConnectError(error);
	}

	/**
	 * Handle disconnection event and clean up pending requests.
	 *
	 * When the connection is lost all outstanding requests are immediately
	 * rejected with a "connection lost" error so callers don't wait forever.
	 * After cleanup the event is propagated to parent classes.
	 *
	 * @param reason - Human-readable reason for the disconnection
	 * @param hasError - Whether the disconnection was caused by an error condition
	 */
	async onDisconnected(reason?: string, hasError = false): Promise<void> {
		// Create a single error object reused for rejecting all pending requests,
		// preserving the disconnect reason as the error message
		const connectionError = new Error(reason || 'Connection lost');

		// Cancel all pending requests: clear their timers and reject the promises
		// so callers receive a prompt error instead of hanging indefinitely
		for (const [, request] of this._pendingRequests) {
			// Clear the per-request timeout handle to avoid a stale timer firing
			if (request.timer) clearTimeout(request.timer);
			// Reject with the connection-lost error
			request.reject(connectionError);
		}

		// Remove all entries now that every promise has been settled
		this._pendingRequests.clear();

		// Propagate to DAPBase / RocketRideClient for additional cleanup
		await super.onDisconnected(reason, hasError);
	}

	/**
	 * Handle received messages from the transport layer.
	 *
	 * Routes incoming messages to the appropriate handler based on their type:
	 * - `response`: matched to a pending request via `request_seq` and resolved
	 * - `event`: forwarded to `onEvent` for user-defined processing
	 * - anything else: logged as an unhandled message type
	 *
	 * @param message - The DAP message received from the server
	 */
	async onReceive(message: DAPMessage): Promise<void> {
		const messageType = message.type;

		if (messageType === 'response') {
			// --- Response path: correlate with the originating request ---

			// The server echoes back the client's sequence number in request_seq
			const requestSeq = message.request_seq;

			if (requestSeq !== undefined && this._pendingRequests.has(requestSeq)) {
				// Retrieve and remove the pending entry atomically to avoid double-resolve
				const pendingRequest = this._pendingRequests.get(requestSeq)!;
				this._pendingRequests.delete(requestSeq);

				// Cancel the per-request timeout now that the response has arrived
				if (pendingRequest.timer) clearTimeout(pendingRequest.timer);

				// Settle the promise with the server's response message
				pendingRequest.resolve(message);
			} else {
				// Response arrived for a request we no longer track (timed out, or spurious)
				this.debugMessage(`Response received for unknown request: ${JSON.stringify(message)}`);
			}
		} else if (messageType === 'event') {
			// --- Event path: server-pushed notifications (status updates, pipeline events) ---
			// Delegate to the virtual onEvent handler (overridden by RocketRideClient)
			await this.onEvent(message);
		} else {
			// --- Unknown message type: log for diagnostics ---
			this.debugMessage(`Unhandled message type: ${JSON.stringify(message)}`);
		}
	}

	/**
	 * Send a request and wait for the corresponding response.
	 *
	 * Assigns a unique sequence number to the message, registers a pending
	 * promise, starts an optional timeout, transmits the message via the
	 * transport, and returns the server's response when it arrives.
	 *
	 * @param message - The DAP message to send (must have `command` and `type` set)
	 * @param timeout - Optional per-request timeout in ms. Overrides the default requestTimeout.
	 * @returns Promise that resolves with the server's response DAPMessage
	 * @throws Error if the message is malformed, the client is not connected, or the request times out
	 */
	async request(message: DAPMessage, timeout?: number): Promise<DAPMessage> {
		// Validate required fields before touching the network
		if (!message.command) {
			this.raiseException(new Error("Request message must include a 'command' field"));
		}

		if (!message.type) {
			this.raiseException(new Error("Request message must include a 'type' field"));
		}

		// Guard: refuse to queue a request when the transport is not connected
		if (!this._transport?.isConnected()) {
			this.raiseException(new Error('Server is not connected'));
		}

		// Assign the next monotonically increasing sequence number to this request
		// so the matching response can be correlated back
		const seq = this.getNextSeq();
		message.seq = seq;

		// Return a promise whose resolve/reject are stored in _pendingRequests
		// and will be called when a matching response arrives (or on timeout/error)
		return new Promise((resolve, reject) => {
			const entry: { resolve: typeof resolve; reject: typeof reject; timer?: ReturnType<typeof setTimeout> } = { resolve, reject };

			// Determine the effective timeout: per-call override takes priority over
			// the instance-level default; undefined means no timeout.
			// Compare against undefined explicitly (rather than truthiness) so that
			// an explicit 0 — e.g. when _dapConnect's budget is already exhausted —
			// produces an immediate timeout rejection instead of hanging forever.
			const effectiveTimeout = timeout ?? this._requestTimeout;
			if (effectiveTimeout !== undefined) {
				// Schedule automatic rejection after the timeout expires
				entry.timer = setTimeout(() => {
					if (this._pendingRequests.has(seq)) {
						// Remove the entry so the response (if it ever arrives) is silently ignored
						this._pendingRequests.delete(seq);
						reject(new Error(`Request timed out after ${effectiveTimeout}ms`));
					}
				}, effectiveTimeout);
			}

			// Register the pending entry indexed by sequence number
			this._pendingRequests.set(seq, entry);

			// Asynchronously transmit the message; on send failure, clean up immediately
			this._send(message).catch((_error) => {
				this.debugMessage(`Clearing request due to error: ${seq}`);

				// If the entry is still present (hasn't already timed out or been resolved),
				// cancel its timer and reject the promise with a send-failure error
				if (this._pendingRequests.has(seq)) {
					const pending = this._pendingRequests.get(seq)!;
					if (pending.timer) clearTimeout(pending.timer);
					this._pendingRequests.delete(seq);
				}
				reject(new Error('Could not send request'));
			});
		});
	}

	/**
	 * Open the transport (WebSocket) without sending any authentication.
	 * Authentication is handled at the RocketRideClient level via login().
	 *
	 * @param timeout - Optional timeout in ms for the WebSocket handshake.
	 * @throws Error if the transport is not configured or the connection times out.
	 */
	async _dapConnect(timeout?: number): Promise<void> {
		if (!this._transport) {
			throw new Error('Transport not configured');
		}
		await this._transport.connect(timeout);
	}

	/**
	 * Close connection to the DAP server and clean up resources.
	 *
	 * Delegates to the transport's `disconnect` method which closes the
	 * underlying socket and triggers `onDisconnected` via the transport callback.
	 */
	async disconnect(): Promise<void> {
		// Request the transport to close the connection gracefully
		if (this._transport) {
			await this._transport.disconnect();
		}
	}

	/**
	 * Check if connected to server.
	 *
	 * Delegates to the transport's connection state; returns false when no
	 * transport is present or the socket is not in the OPEN state.
	 */
	isConnected(): boolean {
		return this._transport?.isConnected() || false;
	}
}
